ethdb/pebble/pebble.go

```diff
@@ -0,0 +1,559 @@
+	writeDelayTime      int64     // total time spent in write stalls
+}
+
+func (d *Database) OnCompactionBegin(info pebble.CompactionInfo) {
+	if d.activeComp == 0 {
+		d.compStartTime = time.Now()
+	}
+	if info.Reason == "read" {
+		atomic.AddInt64(&d.seekCompCount, 1)
+	}
+
+	for _, level := range info.Input {
+		if level.Level == 0 {
+			atomic.AddUint32(&d.level0Comp, 1)
+		} else {
+			atomic.AddUint32(&d.nonLevel0Comp, 1)
+		}
+	}
+	d.activeComp++
+}
+
+func (d *Database) OnCompactionEnd(info pebble.CompactionInfo) {
+	if d.activeComp == 1 {
+		atomic.AddInt64(&d.compTime, int64(time.Since(d.compStartTime)))
+	} else if d.activeComp == 0 {
+		panic("should not happen")
+	}
+
+	d.activeComp--
+}
+
+func (d *Database) OnWriteStallBegin(b pebble.WriteStallBeginInfo) {
+	d.writeDelayStartTime = time.Now()
+}
+
+func (d *Database) OnWriteStallEnd() {
+	atomic.AddInt64(&d.writeDelayTime, int64(time.Since(d.writeDelayStartTime)))
+}
+
+// New returns a wrapped LevelDB object. The namespace is the prefix that the
+// metrics reporting should use for surfacing internal stats.
+func New(file string, cache int, handles int, namespace string, readonly bool) (*Database, error) {
+	var pebbleDb *Database
+	// Ensure we have some minimal caching and file guarantees
+	if cache < minCache {
+		cache = minCache
+	}
+	if handles < minHandles {
+		handles = minHandles
+	}
+	logger := log.New("database", file)
+	logger.Info("Allocated cache and file handles", "cache", common.StorageSize(cache*1024*1024), "handles", handles)
+
+	eventListener := pebble.EventListener{
+		CompactionBegin: func(info pebble.CompactionInfo) {
+			pebbleDb.OnCompactionBegin(info)
+		},
+		CompactionEnd: func(info pebble.CompactionInfo) {
+			pebbleDb.OnCompactionEnd(info)
+		},
+		WriteStallBegin: func(info pebble.WriteStallBeginInfo) {
+			pebbleDb.OnWriteStallBegin(info)
+		},
+		WriteStallEnd: func() {
+			pebbleDb.OnWriteStallEnd()
+		},
+	}
+	// Open the db and recover any potential corruptions
+	db, err := pebble.Open(file, &pebble.Options{
+		// Pebble has a single combined cache area and the write
+		// buffers are taken from this too. Assign all available
+		// memory allowance for cache.
+		Cache:        pebble.NewCache(int64(cache * 1024 * 1024)),
+		MaxOpenFiles: handles,
+		// The size of memory table(as well as the write buffer).
+		// Note, there may have more than two memory tables in the system.
+		// MemTableStopWritesThreshold can be configured to avoid the memory abuse.
+		MemTableSize: cache * 1024 * 1024 / 4,
+		// The default compaction concurrency(1 thread),
+		// Here use all available CPUs for faster compaction.
+		MaxConcurrentCompactions: func() int { return runtime.NumCPU()},
+		// Per-level options. Options for at least one level must be specified. The
+		// options for the last level are used for all subsequent levels.
+		Levels: []pebble.LevelOptions{
+			{TargetFileSize: 2 * 1024 * 1024, FilterPolicy: bloom.FilterPolicy(10)},
```

Comment:
Did we every try with any larger files in pebble?

Reply:
I didn't.

Reply:
I didn't.  Maybe @rjl493456442 did at some point?

Reply:
Yes, I did. As far as I remember it's the same as with LevelDB, compaction makes it go boom.